test/e2e/rule_test.go

```diff
@@ -556,6 +560,141 @@ func TestRule(t *testing.T) {
+						Files:           []string{filepath.Join(e2e.ContainerSharedDir, amTargetsSubDir, "*.yaml")},
+						RefreshInterval: model.Duration(time.Second),
+					},
+				},
+				StaticAddresses: []string{
+					am.NetworkHTTPEndpoint(),
+				},
+				Scheme: "http",
+			},
+			Timeout:    model.Duration(time.Second),
+			APIVersion: alert.APIv1,
+		},
+	}, []query.Config{
+		{
+			EndpointsConfig: http_util.EndpointsConfig{
+				// We test Statically Addressed queries in other tests. Focus on FileSD here.
+				FileSDConfigs: []http_util.FileSDConfig{
+					{
+						// FileSD which will be used to register discover dynamically q.
+						Files:           []string{filepath.Join(e2e.ContainerSharedDir, queryTargetsSubDir, "*.yaml")},
+						RefreshInterval: model.Duration(time.Second),
+					},
+				},
+				Scheme: "http",
+			},
+		},
+	}, true,  remotewrite.Config{
+		Name: "ruler-rw-receivers",
+		RemoteStore: &config.RemoteWriteConfig{
+			URL: &commoncfg.URL{URL: rwURL},
+			Name: "thanos-receiver",
+		},
+	})
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(r))
+
+	writeTargets(t, filepath.Join(s.SharedDir(), queryTargetsSubDir, "targets.yaml"), querier.NetworkHTTPEndpoint())
+	writeTargets(t, filepath.Join(s.SharedDir(), amTargetsSubDir, "targets.yaml"), am.NetworkHTTPEndpoint())
+
+	t.Run("inject samples into receiver to reset its StoreAPI MinTime", func(t *testing.T) {
+		// inject data into receiver to reset its minTime (so it doesn't get filtered out by store)
+		// the sample is injected through a prometheus instance that remote_writes samples into the receiver node
+		prom, _, err := e2ethanos.NewPrometheus(s.SharedDir(), "1", defaultPromConfig("prom", 0, e2ethanos.RemoteWriteEndpoint(receiver.NetworkEndpoint(8081)), ""), e2ethanos.DefaultPrometheusImage())
+		testutil.Ok(t, err)
+		testutil.Ok(t, s.StartAndWaitReady(prom))
+
+		queryAndAssertSeries(t, ctx, querier.HTTPEndpoint(), queryUpWithoutInstance, promclient.QueryOptions{
+			Deduplicate: false,
+		}, []model.Metric{
+			{
+				"job":"myself",
+				"prometheus": "prom",
+				"receive": "1",
+				"replica": "0",
+				"tenant_id": "default-tenant",
+			},
+		})
+	})
+
+	t.Run("query can contact from receiver", func(t *testing.T) {
+		testAbsentQuery := "absent(nonexistent{job='thanos-receive'})"
+		queryAndAssertSeries(t, ctx, querier.HTTPEndpoint(), testAbsentQuery, promclient.QueryOptions{
+			Deduplicate: false,
+		}, []model.Metric{
+			{
+				"job": "thanos-receive",
+			},
+		})
+	})
+
+	t.Run("can fetch remote-written samples from receiver", func(t *testing.T) {
+		testRecordedSamples := "test_absent_metric"
+		queryAndAssertSeries(t, ctx, querier.HTTPEndpoint(), testRecordedSamples, promclient.QueryOptions{
+			Deduplicate: false,
+		}, []model.Metric{
+			{
+				"__name__": "test_absent_metric",
+				"job":"thanos-receive",
+				"receive": "1",
+				"tenant_id": "default-tenant",
+			},
+		})
+	})
```

Comment:
Since we have Alertmanager, maybe we want to create alert and check if the alert was triggered? (But not sure if we want to test that if we already tested this for non stateless ruler)

Reply:
Turns out that adding alerting rules stops the ruler from starting up (causing the test to fail). Here's the relevant log message:
```
19:39:32 rule-1: panic: runtime error: invalid memory address or nil pointer dereference
19:39:32 rule-1: [signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x8efd92]
19:39:32 rule-1: goroutine 354 [running]:
19:39:32 rule-1: github.com/prometheus/prometheus/storage.(*genericSeriesSetAdapter).Next(0xc000a13b90, 0xc000a42e10)
19:39:32 rule-1: <autogenerated>:1 +0x32
19:39:32 rule-1: github.com/prometheus/prometheus/storage.newGenericMergeSeriesSet(0xc000a849e0, 0x2, 0x2, 0xc000a13b60, 0xc000a13bd0, 0xc000cb1960)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/storage/merge.go:319 +0x16c
19:39:32 rule-1: github.com/prometheus/prometheus/storage.(*mergeGenericQuerier).Select.func4(0xc000a849c0, 0x3, 0x4)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/storage/merge.go:143 +0x50
19:39:32 rule-1: github.com/prometheus/prometheus/storage.(*lazyGenericSeriesSet).Next(0xc000a42de0, 0xc000a42de0)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/storage/lazy.go:28 +0x3b
19:39:32 rule-1: github.com/prometheus/prometheus/rules.(*Group).RestoreForState.func2(0xc0006ce0c0)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/rules/manager.go:751 +0x172
19:39:32 rule-1: github.com/prometheus/prometheus/rules.(*AlertingRule).ForEachActiveAlert(0xc000a87ba0, 0xc000cb1d00)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/rules/alerting.go:477 +0xed
19:39:32 rule-1: github.com/prometheus/prometheus/rules.(*Group).RestoreForState(0xc0008620e0, 0xc028b4f937e8d55f, 0x790615b7, 0x2f1eb60)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/rules/manager.go:736 +0x377
19:39:32 rule-1: github.com/prometheus/prometheus/rules.(*Group).run(0xc0008620e0, 0x2115af0, 0xc0009682d0)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/rules/manager.go:400 +0xb98
19:39:32 rule-1: github.com/prometheus/prometheus/rules.(*Manager).Update.func1(0xc00059e200, 0x0, 0xc000041280, 0xc000142780, 0xc0008620e0)
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/rules/manager.go:988 +0x8d
19:39:32 rule-1: created by github.com/prometheus/prometheus/rules.(*Manager).Update
19:39:32 rule-1: /home/michael/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20210519120135-d95b0972505f/rules/manager.go:978 +0x5a5
```

Not sure, but I'm guessing it's because of how the WAL storage implements the Querier interface (https://github.com/idoqo/thanos/blob/7d6095cf27dd8e827db174bcb93ccda722c9ea6e/pkg/rules/remotewrite/wal.go#L678)

Reply:
Thatrs important bug that we have to solve. Looks like alerting was essentially checking WAL for something, which is not possible now, - we need to track what. I think this is the `RestoreForState` function - we probably have to ask querier for it or something - this might need some thoughts.

Great your test caught it!